File: packages/KNIRVCHAIN/internal/indexing/hnsw.go
package indexing

import (
	"container/heap"
	"math"
	"math/rand"
	"sync"
	"time"

	"github.com/google/uuid"
)

var ErrDimensionMismatch = &IndexError{"dimension mismatch"}

type IndexError struct {
	msg string
}

func (e *IndexError) Error() string {
	return e.msg
}

type HNSWNode struct {
	ID          uuid.UUID
	Vector      []float32
	Connections map[int][]uuid.UUID
	Level       int
}

type HNSWIndex struct {
	dimension      int
	M              int
	Mmax           int
	Mmax0          int
	efConstruction int
	ef             int
	ml             float64
	nodes          map[uuid.UUID]*HNSWNode
	entryPoint     *HNSWNode
	mu             sync.RWMutex
	randSource     *rand.Rand
}

func NewHNSWIndex(dimension, m, efConstruction int) *HNSWIndex {
	return &HNSWIndex{
		dimension:      dimension,
		M:              m,
		Mmax:           m,
		Mmax0:          m * 2,
		efConstruction: efConstruction,
		ef:             efConstruction,
		ml:             1.0 / math.Log(2.0),
		nodes:          make(map[uuid.UUID]*HNSWNode),
		entryPoint:     nil,
		randSource:     rand.New(rand.NewSource(time.Now().UnixNano())),
	}
}

func (h *HNSWIndex) SetEf(ef int) {
	h.mu.Lock()
	defer h.mu.Unlock()
	h.ef = ef
}

func (h *HNSWIndex) Add(id uuid.UUID, vector []float32) error {
	h.mu.Lock()
	defer h.mu.Unlock()

	if len(vector) != h.dimension {
		return ErrDimensionMismatch
	}

	level := h.randomLevel()

	node := &HNSWNode{
		ID:          id,
		Vector:      vector,
		Connections: make(map[int][]uuid.UUID),
		Level:       level,
	}

	for l := 0; l <= level; l++ {
		node.Connections[l] = make([]uuid.UUID, 0)
	}

	if h.entryPoint == nil {
		h.entryPoint = node
		h.nodes[id] = node
		return nil
	}

	ep := []uuid.UUID{h.entryPoint.ID}

	for lc := h.entryPoint.Level; lc > level; lc-- {
		ep = h.searchLayer(vector, ep, 1, lc)
	}

	for lc := level; lc >= 0; lc-- {
		candidates := h.searchLayer(vector, ep, h.efConstruction, lc)
		ep = h.selectNeighbors(vector, candidates, h.efConstruction, lc)

		for _, candidateID := range ep {
			if candidateID == id {
				continue
			}
			node.Connections[lc] = append(node.Connections[lc], candidateID)
			if candidate, ok := h.nodes[candidateID]; ok {
				candidate.Connections[lc] = append(candidate.Connections[lc], id)
			}
		}

		if lc == 0 {
			break
		}
	}

	h.nodes[id] = node
	if level > h.entryPoint.Level {
		h.entryPoint = node
	}

	return nil
}

func (h *HNSWIndex) randomLevel() int {
	l := -int(math.Log(h.randSource.Float64()) * h.ml)
	if l < 0 {
		l = 0
	}
	return l
}

func (h *HNSWIndex) searchLayer(query []float32, entryPoints []uuid.UUID, ef, layer int) []uuid.UUID {
	var visited = make(map[uuid.UUID]bool)
	for _, ep := range entryPoints {
		visited[ep] = true
	}

	candidates := &priorityQueue{}
	heap.Init(candidates)

	for _, epID := range entryPoints {
		if node, ok := h.nodes[epID]; ok {
			dist := h.distance(query, node.Vector)
			heap.Push(candidates, &pqItem{id: epID, dist: dist})
		}
	}

	result := make([]uuid.UUID, 0)
	resultDists := make(map[uuid.UUID]float32)

	for candidates.Len() > 0 {
		current := heap.Pop(candidates).(*pqItem)
		result = append(result, current.id)
		resultDists[current.id] = current.dist

		if len(result) > ef {
			break
		}

		if node, ok := h.nodes[current.id]; ok {
			for _, neighborID := range node.Connections[layer] {
				if !visited[neighborID] {
					visited[neighborID] = true
					if neighbor, ok := h.nodes[neighborID]; ok {
						dist := h.distance(query, neighbor.Vector)
						if len(result) <= ef || dist <= resultDists[result[len(result)-1]] {
							heap.Push(candidates, &pqItem{id: neighborID, dist: dist})
						}
					}
				}
			}
		}
	}

	return result
}

func (h *HNSWIndex) selectNeighbors(query []float32, candidates []uuid.UUID, m, layer int) []uuid.UUID {
	if len(candidates) == 0 {
		return nil
	}

	neighbors := make([]uuid.UUID, 0)

	if layer == 0 {
		pq := &priorityQueue{}
		heap.Init(pq)

		for _, candidateID := range candidates {
			if node, ok := h.nodes[candidateID]; ok {
				dist := h.distance(query, node.Vector)
				heap.Push(pq, &pqItem{id: candidateID, dist: dist})
			}
		}

		for pq.Len() > 0 && len(neighbors) < h.Mmax0 {
			item := heap.Pop(pq).(*pqItem)
			neighbors = append(neighbors, item.id)
		}
	} else {
		pq := &priorityQueue{}
		heap.Init(pq)

		for _, candidateID := range candidates {
			if node, ok := h.nodes[candidateID]; ok {
				dist := h.distance(query, node.Vector)
				heap.Push(pq, &pqItem{id: candidateID, dist: dist})
			}
		}

		for pq.Len() > 0 && len(neighbors) < h.Mmax {
			item := heap.Pop(pq).(*pqItem)
			neighbors = append(neighbors, item.id)
		}
	}

	return neighbors
}

func (h *HNSWIndex) Search(query []float32, k int) ([]uuid.UUID, error) {
	h.mu.RLock()
	defer h.mu.RUnlock()

	if len(query) != h.dimension {
		return nil, ErrDimensionMismatch
	}

	if h.entryPoint == nil {
		return []uuid.UUID{}, nil
	}

	ep := []uuid.UUID{h.entryPoint.ID}

	for lc := h.entryPoint.Level; lc > 0; lc-- {
		ep = h.searchLayer(query, ep, 1, lc)
	}

	candidates := h.searchLayer(query, ep, h.ef, 0)
	results := h.selectNeighbors(query, candidates, k, 0)

	if len(results) > k {
		results = results[:k]
	}

	return results, nil
}

func (h *HNSWIndex) Remove(id uuid.UUID) error {
	h.mu.Lock()
	defer h.mu.Unlock()

	node, ok := h.nodes[id]
	if !ok {
		return nil
	}

	for layer := 0; layer <= node.Level; layer++ {
		for _, neighborID := range node.Connections[layer] {
			if neighbor, ok := h.nodes[neighborID]; ok {
				newConns := make([]uuid.UUID, 0)
				for _, connID := range neighbor.Connections[layer] {
					if connID != id {
						newConns = append(newConns, connID)
					}
				}
				neighbor.Connections[layer] = newConns
			}
		}
	}

	delete(h.nodes, id)

	if h.entryPoint != nil && h.entryPoint.ID == id {
		h.entryPoint = nil
		for _, n := range h.nodes {
			if h.entryPoint == nil || n.Level > h.entryPoint.Level {
				h.entryPoint = n
			}
		}
	}

	return nil
}

func (h *HNSWIndex) distance(a, b []float32) float32 {
	var sum float32
	for i := range a {
		d := a[i] - b[i]
		sum += d * d
	}
	return sum
}

type pqItem struct {
	id   uuid.UUID
	dist float32
}

type priorityQueue []*pqItem

func (pq priorityQueue) Len() int           { return len(pq) }
func (pq priorityQueue) Less(i, j int) bool { return pq[i].dist < pq[j].dist }
func (pq priorityQueue) Swap(i, j int)      { pq[i], pq[j] = pq[j], pq[i] }
func (pq *priorityQueue) Push(x interface{}) {
	*pq = append(*pq, x.(*pqItem))
}
func (pq *priorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	*pq = old[0 : n-1]
	return item
}

func timeNow() interface{ UnixNano() int64 } {
	return &mockTime{}
}

type mockTime struct{}

func (m *mockTime) UnixNano() int64 {
	return rand.Int63()
}

File: packages/KNIRVBASE/rust/src/hnsw.rs
use rand::Rng;
use serde::{Deserialize, Serialize};
use std::cmp::Ordering;
use std::collections::{BinaryHeap, HashMap};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HNSWNode {
    pub id: Uuid,
    pub vector: Vec<f32>,
    pub connections: HashMap<i32, Vec<Uuid>>,
    pub level: i32,
}

#[derive(Debug, Clone)]
pub struct HNSWIndex {
    dimension: usize,
    m: usize,
    _mmax: usize,
    mmax0: usize,
    ef_construction: usize,
    ef: usize,
    _ml: f64,
    nodes: HashMap<Uuid, HNSWNode>,
    entry_point: Option<Uuid>,
}

#[derive(Debug)]
pub struct HNSWError {
    pub msg: String,
}

impl HNSWError {
    pub fn new(msg: &str) -> Self {
        Self {
            msg: msg.to_string(),
        }
    }
}

impl std::fmt::Display for HNSWError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.msg)
    }
}

impl std::error::Error for HNSWError {}

#[derive(Clone)]
struct DistanceItem {
    id: Uuid,
    distance: f64,
}

impl PartialEq for DistanceItem {
    fn eq(&self, other: &Self) -> bool {
        self.distance == other.distance
    }
}

impl Eq for DistanceItem {}

impl PartialOrd for DistanceItem {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for DistanceItem {
    fn cmp(&self, other: &Self) -> Ordering {
        self.distance
            .partial_cmp(&other.distance)
            .unwrap_or(Ordering::Equal)
    }
}

impl HNSWIndex {
    pub fn new(dimension: usize, m: usize, ef_construction: usize) -> Self {
        Self {
            dimension,
            m,
            _mmax: m,
            mmax0: m * 2,
            ef_construction,
            ef: ef_construction,
            _ml: 1.0 / (2.0_f64.ln()),
            nodes: HashMap::new(),
            entry_point: None,
        }
    }

    pub fn set_ef(&mut self, ef: usize) {
        self.ef = ef;
    }

    pub fn add(&mut self, id: Uuid, vector: Vec<f32>) -> Result<(), HNSWError> {
        if vector.len() != self.dimension {
            return Err(HNSWError::new("dimension mismatch"));
        }

        let level = self.random_level();

        let mut node = HNSWNode {
            id,
            vector,
            connections: HashMap::new(),
            level,
        };

        for l in 0..=level {
            node.connections.insert(l, Vec::new());
        }

        if self.entry_point.is_none() {
            self.entry_point = Some(id);
            self.nodes.insert(id, node);
            return Ok(());
        }

        let entry_point_id = self.entry_point.unwrap();
        let mut ep = vec![entry_point_id];

        if let Some(entry) = self.nodes.get(&entry_point_id) {
            for lc in (level + 1)..=entry.level {
                ep = self.search_layer(&node.vector, &ep, 1, lc);
            }
        }

        for lc in (0..=level).rev() {
            let m = if lc == 0 { self.mmax0 } else { self.m };
            let candidates = self.search_layer(&node.vector, &ep, self.ef_construction, lc);
            let neighbors = self.select_neighbors(&node.vector, &candidates, m);

            for &neighbor_id in &neighbors {
                self.connect(id, neighbor_id, lc);
                self.connect(neighbor_id, id, lc);

                if let Some(neighbor_node) = self.nodes.get(&neighbor_id) {
                    let conns = neighbor_node
                        .connections
                        .get(&lc)
                        .map(|c| c.len())
                        .unwrap_or(0);
                    if conns > m {
                        self.prune_connections(neighbor_id, lc, m);
                    }
                }
            }

            ep = candidates;
        }

        if level
            > self
                .nodes
                .get(&self.entry_point.unwrap())
                .map(|n| n.level)
                .unwrap_or(-1)
        {
            self.entry_point = Some(id);
        }

        self.nodes.insert(id, node);
        Ok(())
    }

    pub fn search(&self, query: &[f32], k: usize) -> Result<Vec<Uuid>, HNSWError> {
        if query.len() != self.dimension {
            return Err(HNSWError::new("dimension mismatch"));
        }

        let entry = match self.entry_point {
            Some(id) => id,
            None => return Ok(Vec::new()),
        };

        let mut ep = if self.nodes.len() <= k * 2 {
            self.nodes.keys().cloned().collect::<Vec<_>>()
        } else {
            vec![entry]
        };

        if let Some(entry_node) = self.nodes.get(&entry) {
            for lc in (1..=entry_node.level).rev() {
                ep = self.search_layer(query, &ep, 1, lc);
            }
        }

        ep = self.search_layer(query, &ep, self.ef.max(k), 0);

        let mut entries: Vec<(Uuid, f64)> = ep
            .iter()
            .filter_map(|id| {
                self.nodes.get(id).map(|node| {
                    let dist = euclidean_distance(query, &node.vector);
                    (*id, dist)
                })
            })
            .collect();

        entries.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap_or(Ordering::Equal));

        Ok(entries.into_iter().take(k).map(|(id, _)| id).collect())
    }

    fn search_layer(
        &self,
        query: &[f32],
        entry_points: &[Uuid],
        ef: usize,
        layer: i32,
    ) -> Vec<Uuid> {
        let mut candidates: BinaryHeap<DistanceItem> = BinaryHeap::new();
        let mut results: BinaryHeap<DistanceItem> = BinaryHeap::new();
        let mut visited: HashMap<Uuid, bool> = HashMap::new();

        for &ep in entry_points {
            if let Some(node) = self.nodes.get(&ep) {
                let dist = euclidean_distance(query, &node.vector);
                candidates.push(DistanceItem {
                    id: ep,
                    distance: dist,
                });
                results.push(DistanceItem {
                    id: ep,
                    distance: -dist,
                });
                visited.insert(ep, true);
            }
        }

        while let Some(current) = candidates.pop() {
            if results.len() >= ef
                && current.distance > -results.peek().map(|r| r.distance).unwrap_or(f64::MAX)
            {
                break;
            }

            if let Some(node) = self.nodes.get(&current.id) {
                if let Some(connections) = node.connections.get(&layer) {
                    for &neighbor_id in connections {
                        if !visited.contains_key(&neighbor_id) {
                            visited.insert(neighbor_id, true);

                            if let Some(neighbor) = self.nodes.get(&neighbor_id) {
                                let dist = euclidean_distance(query, &neighbor.vector);

                                if results.len() < ef
                                    || dist
                                        < -results.peek().map(|r| r.distance).unwrap_or(f64::MAX)
                                {
                                    candidates.push(DistanceItem {
                                        id: neighbor_id,
                                        distance: dist,
                                    });
                                    results.push(DistanceItem {
                                        id: neighbor_id,
                                        distance: -dist,
                                    });

                                    if results.len() > ef {
                                        results.pop();
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }

        let mut items: Vec<_> = results.into_vec();
        items.reverse();
        items.into_iter().map(|item| item.id).collect()
    }

    fn select_neighbors(&self, query: &[f32], candidates: &[Uuid], m: usize) -> Vec<Uuid> {
        if candidates.len() <= m {
            return candidates.to_vec();
        }

        let mut cands: Vec<(Uuid, f64)> = candidates
            .iter()
            .filter_map(|id| {
                self.nodes.get(id).map(|node| {
                    let dist = euclidean_distance(query, &node.vector);
                    (*id, dist)
                })
            })
            .collect();

        cands.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap_or(Ordering::Equal));

        cands.into_iter().take(m).map(|(id, _)| id).collect()
    }

    fn connect(&mut self, from: Uuid, to: Uuid, layer: i32) {
        if let Some(node) = self.nodes.get_mut(&from) {
            if let Some(connections) = node.connections.get_mut(&layer) {
                if !connections.contains(&to) {
                    connections.push(to);
                }
            }
        }
    }

    fn prune_connections(&mut self, node_id: Uuid, layer: i32, m: usize) {
        let node = match self.nodes.get(&node_id) {
            Some(n) => n,
            None => return,
        };

        let connections = match node.connections.get(&layer) {
            Some(c) => c.clone(),
            None => return,
        };

        if connections.len() <= m {
            return;
        }

        let mut distances: Vec<(Uuid, f64)> = connections
            .iter()
            .filter_map(|conn_id| {
                self.nodes.get(conn_id).map(|conn_node| {
                    let dist = euclidean_distance(&node.vector, &conn_node.vector);
                    (*conn_id, dist)
                })
            })
            .collect();

        distances.sort_by(|a, b| a.1.partial_cmp(&b.1).unwrap_or(Ordering::Equal));

        let new_connections: Vec<Uuid> = distances.into_iter().take(m).map(|(id, _)| id).collect();

        if let Some(node) = self.nodes.get_mut(&node_id) {
            node.connections.insert(layer, new_connections);
        }
    }

    fn random_level(&self) -> i32 {
        let mut rng = rand::thread_rng();
        let mut level = 0i32;
        while rng.gen::<f64>() < 0.5 && level < 16 {
            level += 1;
        }
        level
    }

    pub fn remove(&mut self, id: Uuid) -> Result<(), HNSWError> {
        let node = match self.nodes.remove(&id) {
            Some(n) => n,
            None => return Err(HNSWError::new("node not found")),
        };

        for layer in 0..=node.level {
            for &neighbor_id in node
                .connections
                .get(&layer)
                .map(|c| c.as_slice())
                .unwrap_or(&[])
            {
                if let Some(neighbor) = self.nodes.get_mut(&neighbor_id) {
                    if let Some(connections) = neighbor.connections.get_mut(&layer) {
                        connections.retain(|&conn| conn != id);
                    }
                }
            }
        }

        if self.entry_point == Some(id) {
            self.entry_point = self.find_new_entry_point();
        }

        Ok(())
    }

    fn find_new_entry_point(&self) -> Option<Uuid> {
        self.nodes
            .values()
            .max_by_key(|node| node.level)
            .map(|node| node.id)
    }

    pub fn size(&self) -> usize {
        self.nodes.len()
    }

    pub fn get_all_nodes(&self) -> Vec<Uuid> {
        self.nodes.keys().cloned().collect()
    }

    pub fn get_node(&self, id: &Uuid) -> Option<&HNSWNode> {
        self.nodes.get(id)
    }
}

fn euclidean_distance(v1: &[f32], v2: &[f32]) -> f64 {
    if v1.len() != v2.len() {
        return f64::MAX;
    }

    let mut sum = 0.0_f64;
    for i in 0..v1.len() {
        let diff = v1[i] as f64 - v2[i] as f64;
        sum += diff * diff;
    }

    sum.sqrt()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_hnsw_add_and_search() {
        let mut index = HNSWIndex::new(4, 16, 200);

        let id1 = Uuid::new_v4();
        let id2 = Uuid::new_v4();
        let id3 = Uuid::new_v4();

        index.add(id1, vec![1.0, 0.0, 0.0, 0.0]).unwrap();
        index.add(id2, vec![0.0, 1.0, 0.0, 0.0]).unwrap();
        index.add(id3, vec![0.9, 0.1, 0.0, 0.0]).unwrap();

        let results = index.search(&[1.0, 0.0, 0.0, 0.0], 2).unwrap();

        assert_eq!(results.len(), 2);
        assert_eq!(results[0], id1);
    }

    #[test]
    fn test_hnsw_remove() {
        let mut index = HNSWIndex::new(4, 16, 200);

        let id1 = Uuid::new_v4();
        let id2 = Uuid::new_v4();

        index.add(id1, vec![1.0, 0.0, 0.0, 0.0]).unwrap();
        index.add(id2, vec![0.0, 1.0, 0.0, 0.0]).unwrap();

        assert_eq!(index.size(), 2);

        index.remove(id1).unwrap();

        assert_eq!(index.size(), 1);
        assert!(index.get_node(&id1).is_none());
        assert!(index.get_node(&id2).is_some());
    }

    #[test]
    fn test_dimension_mismatch() {
        let mut index = HNSWIndex::new(4, 16, 200);

        let result = index.add(Uuid::new_v4(), vec![1.0, 0.0, 0.0]);
        assert!(result.is_err());
    }
}
